Java/core/src/main/java/com/amazon/randomcutforest/imputation/Summarizer.java

```diff
@@ -0,0 +1,371 @@
+
+    /**
+     * adds a center corresponding to point number index in the list, provided there
+     * is already no other center with the same coordinates (or rather, ruling out
+     * distance 0 from other centers)
+     * 
+     * @param points  a list of points
+     * @param centers a list of current centers
+     * @param dist    distance function
+     * @param index   the index of the point to be added as a center
+     */
+    public static void addCenter(ArrayList<ProjectedPoint> points, ArrayList<Center> centers,
+            BiFunction<float[], float[], Double> dist, int index) {
+        double minDist = Double.MAX_VALUE;
+        for (int i = 0; i < centers.size(); i++) {
+            double t = dist.apply(centers.get(i).coordinate, points.get(index).coordinate);
+            minDist = min(t, minDist);
+        }
+        if (minDist > 0.0) {
+            centers.add(new Center(points.get(index).coordinate, points.get(index).weight));
+        }
+    }
+
+    /**
+     * picks an item such that the prefix sum of the weight of the projectedPoints
+     * is at most wt
+     * 
+     * @param points a list of projected points (points with weight)
+     * @param wt     a selection parameter
+     * @return the index of the point for which the condition is satisfied (or the
+     *         last index)
+     */
+    public static int pick(ArrayList<ProjectedPoint> points, double wt) {
+        int position = 0;
+        double running = wt;
+        for (int i = 0; i < points.size(); i++) {
+            position = i;
+            if (running - points.get(i).weight <= 0.0) {
+                break;
+            } else {
+                running -= points.get(i).weight;
+            }
+        }
+        return position;
+    }
+
+    /**
+     * creates a sample of points based on an input list; it first adds all points
+     * with large weights (0.1%, controlled by LENGTH_BOUND), and in the remainder
+     * performs a standard MonteCarlo sampling (adjusting the weight of the sampled
+     * elements keeping the total approximately the same) The numbers are chosen
+     * such that the total number would be not much more than 10 * LENGTH_BOUND Note
+     * that performing a sampling before grouping can be beneficial both in speed as
+     * well as denoising see https://en.wikipedia.org/wiki/CURE_algorithm and
+     * https://en.wikipedia.org/wiki/Data_stream_clustering the current sampling can
+     * be used with intended spherical clusters as in here, or in future
+     * non-spherical clusters as well
+     *
+     * @param points      list of points with weights
+     * @param seed        a randomseed to control/reproduce the sampling
+     * @param totalWeight the sum of weights of the points (this can be approximate)
+     * @return a sample of points (preserving the same order)
+     */
+
+    static ArrayList<ProjectedPoint> createSample(ProjectedPoint[] points, long seed, double totalWeight) {
```

Comment:
Maybe we should move this to a utility class and give it a name indicating that it's a specific sampling algorithm and maybe the the first one people would think of if you handed them a list of points and weights and asked them to create a sample.